source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,459 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/cal/private/rsa.h>
+
+#include <aws/cal/cal.h>
+#include <aws/cal/private/der.h>
+#include <aws/common/encoding.h>
+
+#include <windows.h>
+
+#include <bcrypt.h>
+
+static BCRYPT_ALG_HANDLE s_rsa_alg = NULL;
+
+static aws_thread_once s_rsa_thread_once = AWS_THREAD_ONCE_STATIC_INIT;
+
+static void s_load_alg_handle(void *user_data) {
+    (void)user_data;
+    /* this function is incredibly slow, LET IT LEAK*/
+    NTSTATUS status = BCryptOpenAlgorithmProvider(&s_rsa_alg, BCRYPT_RSA_ALGORITHM, MS_PRIMITIVE_PROVIDER, 0);
+    AWS_FATAL_ASSERT(s_rsa_alg && "BCryptOpenAlgorithmProvider() failed");
+    AWS_FATAL_ASSERT(BCRYPT_SUCCESS(status));
+}
+
+struct bcrypt_rsa_key_pair {
+    struct aws_rsa_key_pair base;
+    BCRYPT_KEY_HANDLE key_handle;
+    struct aws_byte_buf key_buf;
+};
+
+static void s_rsa_destroy_key(struct aws_rsa_key_pair *key_pair) {
+    if (key_pair == NULL) {
+        return;
+    }
+
+    struct bcrypt_rsa_key_pair *rsa_key = key_pair->impl;
+
+    if (rsa_key->key_handle) {
+        BCryptDestroyKey(rsa_key->key_handle);
+    }
+    aws_byte_buf_clean_up_secure(&rsa_key->key_buf);
+
+    aws_mem_release(key_pair->allocator, rsa_key);
+}
+
+int s_is_not_supported_enc_algo(enum aws_rsa_encryption_algorithm algorithm) {
+    if (algorithm != AWS_CAL_RSA_ENCRYPTION_PKCS1_5 && algorithm != AWS_CAL_RSA_ENCRYPTION_OAEP_SHA256 &&
+        algorithm != AWS_CAL_RSA_ENCRYPTION_OAEP_SHA512) {
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
```

Comment:
NIT: I typically find the pattern of raising in these to get me into trouble. It's gonna be funny if you copied this from my ecc code, but regardless in hindsight, raising exactly where the check is being performed, (line 62) seems to be less prone for error for me.